Approver-led governance for Kubevirt

jberkus opened the pull request:
Creates a GOVERNANCE.md file which outlines a simple self-selecting-council governance format.

This document is based on the CNCF Maintainer governance template.

Updated 6/26, Assumptions made in the new version per @fabiand:

1. ~~"Approver" in our membership docs is equivalent to "Maintainer".~~ Maintainers become a selected position, instead of the group of Approvers to kubevirt/kubevirt
2. CoC violations will be initially reported to the approvers before being escalated to the CNCF.

Unresolved Issues:

1. We need to clean up membership_policy.md so that it reflects our actual process of contributor advancement, particularly the ability to become Approver in a subproject, and that Approver means specifically root-level approver.
2. We also need a private mailing list for the collected Maintainers, as a location to discuss security escalations.

Fixes #97

Signed-off-by: Josh Berkus <josh@agliodbs.com>

mazzystr commented:
/ok-to-test

fabiand commented:
@jberkus the file is empty.

Also adding
/hold
to get reviews from at least @davidvossel @vladikr @rmohr @mhenriks @phoracek

jberkus commented:
Had a meeting with Fabian and David, text has been heavily revised based on that discussion. Please review and comment!

EdDev commented:
> Assumptions made:
> 
>     1. "Approver" in our membership docs is equivalent to "Maintainer".

The template suggested here points to a single list of maintainers which are suppose to govern the whole kubevirt ecosystem.
But the original approval membership is different per how I perceive it.
The membership doc provides a guideline on how each repo should manage its contributors (except for the organization member which is not per repo). Therefore, there cannot be one maintainers list in one specific place, it is distributed across the repos.
In addition, the maintainers of individual repos are diverse, some are taking active part of many community activities and some are more focused or maintaining the correctness and quality of the code (many times with a SIG specific focus).
The description of a maintainer in this text is different, it requires much more from a maintainer.

> Unresolved Issues:
> 
>     1. We need to clean up membership_policy.md so that it reflects our actual process of contributor advancement, particularly the ability to become Approver in a subproject, and that Approver means specifically root-level approver.

Could you please clarify why it is not currently reflecting the actual process?

Regarding the root-level approvers, I think it is perfectly fine to nominate with the same "rules" SIG approvals. The reality is that in large repos (e.g. kubevirt), the code is not structured in a way that this is practical to enforce it. Therefore, an approver is added to root, but hopefully it is clear that if the nomination was given to maintain a specific SIG, the boundaries will be kept most of the times.

What is not cleared to me in the current text, is the maintainer definition. Is it over the organization? On a per-repo?

davidvossel commented:
> > Assumptions made:
> > ```
> > 1. "Approver" in our membership docs is equivalent to "Maintainer".
> > ```
> 
> The template suggested here points to a single list of maintainers which are suppose to govern the whole kubevirt ecosystem.
> But the original approval membership is different per how I perceive it.
> The membership doc provides a guideline on how each repo should manage its contributors (except for the organization member which is not per repo). Therefore, there cannot be one maintainers list in one specific place, it is distributed across the repos.
> In addition, the maintainers of individual repos are diverse, some are taking active part of many community activities and some are more focused or maintaining the correctness and quality of the code (many times with a SIG specific focus).
> The description of a maintainer in this text is different, it requires much more from a maintainer.

saying an "approver" is equivalent to a KubeVirt CNCF Project Maintainer is seems inaccurate.

While it's likely that a CNCF maintainer has approver rights for the kubevirt/kubevirt repo, my understanding is that these two concepts are orthogonal to one another.

jberkus commented:
@EdDev @davidvossel On 6/26, I changed the actual text to reflect a different understanding of Approver and Maintainer based on the convo with David and Fabian.  However, I just changed the PR comment now to reflect the new text.  Please re-examine, thanks.

davidvossel commented:
> However, I just changed the PR comment now to reflect the new text. Please re-examine, thanks.

thanks, looks accurate now

dankenigsberg commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,106 @@
+individual repository level, and provide strategic guidance for the project
+overall.
+
+The current maintainers can be found in [MAINTAINERS](./MAINTAINERS.md).  
```
Shouldn't this PR include this file, even if empty? Though I suspect we'd better initialize it with the list of people who actually do this work these days.

Instead of a new file, how about adding a `maintainers` alias into the existing OWNERS file? The function of these files would be very similar, and sure to confuse me.

jberkus replied:
Dan: if you read this governance proposal, it specifically separates OWNERS for Maintainers.  That's really the meat of the proposal.  So adding the separate file instead.

dankenigsberg commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,106 @@
+  * Do they understand how the project works (policies, processes, etc)?
+  * Are they willing to take on the additional duties of a maintainer?
+
+A candidate must be proposed by an existing maintainer by filing an issue in the
```
How about, instead, doing having the proposal as a PR to the MAINTAINERS.md file?

jberkus replied:
that's fine, too

dankenigsberg commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,106 @@
+A candidate must be proposed by an existing maintainer by filing an issue in the
+[Community Repo](https://github.com/kubevirt/community). A simple majority vote 
+of +1s from existing Maintainers approves the application. Approved maintainers
+will be added to the private maintainer mailing list.
```
can we create this list and give a URL to it here?

jberkus replied:
Apparently it already exists, added link.

dankenigsberg commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,106 @@
+or Code of Conduct violations.  Such meetings should be scheduled by any
+Maintainer on receipt of a security issue or CoC report.  All current Maintainers
+must be invited to such closed meetings, except for any Maintainer who is
+accused of a CoC violation.
```
is there a quorum for these closed meetings? Does a meeting count if only one Maintainer shows up?

jberkus replied:
this doesn't supercede the requirement for a majority of maintainers to approve *anything*.  So you'd need at least 51% of the maintainers to show up.

dankenigsberg commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,106 @@
+
+Maintainers may also be demoted at any time for one of the following reasons:
+
+* Inactivity, including 6 months of more of non-participation or non-communication,
```
```suggestion
* Inactivity, including 6 months or more of non-participation or non-communication,
```

jberkus commented:
@dankenigsberg  That is the list we have on file with the CNCF, with the exception of Pep, whom I removed.

Supposedly Chris filed an update with them shortly before he went on medical leave, but I don't know what happened after that.

fabiand commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,107 @@
+follow through to fix issues.
+
+A maintainer is a contributor to the project's success and a citizen helping
+the project succeed.
```
Great wording!

fabiand commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,107 @@
+[Community Repo](https://github.com/kubevirt/community) against the MAINTAINERS.md file. 
+A simple majority vote of +1s from existing Maintainers approves the application. 
+Approved maintainers will be added to the [private maintainer mailing list](mailto:cncf-kubevirt-maintainers@lists.cncf.io).
+
```
Should we have a upper limit of maintainers?

jberkus replied:
You can, but I'm not sure you need one?  

The idea is that everyone who is consistently on the list of "if we're making a major project-wide change we'll have to ping this person" will be a Maintainer.  And if you have to solicit that person anyway, then it's actually easier to have them on the maintainers list.  If that ends up being a large group of people eventually, that means we have a lot of folks who are very involved in Kubevirt's future, which is hardly a bad thing.

fabiand commented on MAINTAINERS.md:
```diff
@@ -0,0 +1,15 @@
+
+| Name | Employer | Responsibilities |
+| ----------- | ------- | -------------------------------- |
+| [Marc Sluiter](https://github.com/slintes) | Red Hat | |
```
Marc is awesome, but hasn't been active recently, let's drop him.

fabiand commented on MAINTAINERS.md:
```diff
@@ -0,0 +1,15 @@
+| [David Vossel](https://github.com/davidvossel) | Red Hat | |
+| [Vladik Romanovsky](https://github.com/vladikr) | Red Hat | |
+| [Roman Mohr](https://github.com/rmohr) | Red Hat | |
+| [Artyom Lukianov](https://github.com/cynepco3hahue) | Red Hat | |
```
Artyom is awesome, but hasn't been active recently, let's drop him.

fabiand commented on MAINTAINERS.md:
```diff
@@ -0,0 +1,15 @@
+| [Vladik Romanovsky](https://github.com/vladikr) | Red Hat | |
+| [Roman Mohr](https://github.com/rmohr) | Red Hat | |
+| [Artyom Lukianov](https://github.com/cynepco3hahue) | Red Hat | |
+| [Sebastian Scheinkman](https://github.com/SchSeba) | Red Hat | |
```
Sebastian is awesome, but hasn't been active recently, let's drop him.

fabiand commented on MAINTAINERS.md:
```diff
@@ -0,0 +1,15 @@
+| [Roman Mohr](https://github.com/rmohr) | Red Hat | |
+| [Artyom Lukianov](https://github.com/cynepco3hahue) | Red Hat | |
+| [Sebastian Scheinkman](https://github.com/SchSeba) | Red Hat | |
+| [Marcin Franczyk](https://github.com/mfranczy) | Red Hat | |
```
Marcin is awesome, but hasn't been active recently, let's drop him.

fabiand commented on MAINTAINERS.md:
```diff
@@ -0,0 +1,15 @@
+| [Marcin Franczyk](https://github.com/mfranczy) | Red Hat | |
+| [Fabian Deutsch](https://github.com/fabiand) | Red Hat | |
+| [Stu Gott](https://github.com/stu-gott) | Red Hat | |
+
```
I'd like to propose to add Ryan Hallisey and potentially Marcus Sorensen to the initial list of contributors.

To me both of them are already meeting the maintainer requirements.

fabiand replied:
@rthallisey @mlsorensen do you have thoughts on this?
Would you be willing to be a maintainer and take these responsibilites associated with it?

davidvossel replied:
+1 from me on these two as cncf maintainers for kubevirt

rmohr replied:
As discussed in the community meeting I would first want to see a sign of life from @mlsorensen that he will be available and responsive if needed. Apart from that, both will work from me.

fabiand replied:
We did not get a reply from Marcus and Ryan here, therefor I'm proposing to drop them.

fabiand reviewed (commented):
The oevral model sounds good, specifically the separation between current kubevirt/kubevirt approvers and maintainers.

I've included a suggestion to drop some maintainers and add some others.

rmohr commented on MAINTAINERS.md:
```diff
@@ -0,0 +1,15 @@
+| [Sebastian Scheinkman](https://github.com/SchSeba) | Red Hat | |
+| [Marcin Franczyk](https://github.com/mfranczy) | Red Hat | |
+| [Fabian Deutsch](https://github.com/fabiand) | Red Hat | |
+| [Stu Gott](https://github.com/stu-gott) | Red Hat | |
```
I want to propose @vasiliy-ul from Suse who drove cgroupv2 for kubevirt and was quite active on bug fixing and is maintaining kubevirt for integration in ranger.

rmohr replied:
@davidvossel @fabiand does that work for you? Like for @mlsorensen I would want to wait for a sign of life before just blindly adding them.

vasiliy-ul replied:
Sure, I would be glad to contribute to KubeVirt in the new role if that was acceptable and fine with everyone. Probably I am still not fully familiar with all the project specifics but I am ready to dedicate additional effort to fill the gaps.

fabiand replied:
Thanks for showing the openess, Vasily. We'll support you and are looking forward to your thoughts and opinions.

jberkus commented:
@rmohr, @fabiand if you can have some kind of "official" vote/notice/whatever, I can make all those changes to the maintainer file, merge it, and tell you how to let the CNCF know.

rmohr commented:
> @rmohr, @fabiand if you can have some kind of "official" vote/notice/whatever, I can make all those changes to the maintainer file, merge it, and tell you how to let the CNCF know.

I see that neither @mlsorensen nor @vasiliy-ul responded yet. We are in vacation time, they are probably both on vacation. I think we need a commitment for either one of those if they want to be added.

jberkus commented:
I have removed the inactive maintainers per the above thread.

Per Roman and Fabian, I have added Ryan and Vasiliy as maintainers. I have not added Marcus per discussion above.

fabiand reviewed (commented):
Please update the reviewer list, then we should be good to go. Wohoo

fabiand commented:
@jberkus I'm not seeing the maintainer changes, did you miss to push a commit?

mazzystr commented:
Please add me (Chris Callegari / ***@***.*** / Red Hat) also.  This
allows me to create tickets in the CNCF Service Desk for project related
items.

On Wed, Sep 1, 2021 at 11:17 AM Fabian Deutsch ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Please update the reviewer list, then we should be good to go. Wohoo
> ------------------------------
>
> In MAINTAINERS.md
> <https://github.com/kubevirt/community/pull/123#discussion_r700455394>:
>
> > @@ -0,0 +1,15 @@
> +The current Maintainers Group for the Kubevirt Project consists of:
> +
> +| Name | Employer | Responsibilities |
> +| ----------- | ------- | -------------------------------- |
> +| [Marc Sluiter](https://github.com/slintes) | Red Hat | |
> +| [David Vossel](https://github.com/davidvossel) | Red Hat | |
> +| [Vladik Romanovsky](https://github.com/vladikr) | Red Hat | |
> +| [Roman Mohr](https://github.com/rmohr) | Red Hat | |
> +| [Artyom Lukianov](https://github.com/cynepco3hahue) | Red Hat | |
> +| [Sebastian Scheinkman](https://github.com/SchSeba) | Red Hat | |
> +| [Marcin Franczyk](https://github.com/mfranczy) | Red Hat | |
> +| [Fabian Deutsch](https://github.com/fabiand) | Red Hat | |
> +| [Stu Gott](https://github.com/stu-gott) | Red Hat | |
> +
>
> We did not get a reply from Marcus and Ryan here, therefor I'm proposing
> to drop them.
> ------------------------------
>
> In MAINTAINERS.md
> <https://github.com/kubevirt/community/pull/123#discussion_r700455015>:
>
> > @@ -0,0 +1,15 @@
> +The current Maintainers Group for the Kubevirt Project consists of:
> +
> +| Name | Employer | Responsibilities |
> +| ----------- | ------- | -------------------------------- |
> +| [Marc Sluiter](https://github.com/slintes) | Red Hat | |
> +| [David Vossel](https://github.com/davidvossel) | Red Hat | |
> +| [Vladik Romanovsky](https://github.com/vladikr) | Red Hat | |
> +| [Roman Mohr](https://github.com/rmohr) | Red Hat | |
> +| [Artyom Lukianov](https://github.com/cynepco3hahue) | Red Hat | |
> +| [Sebastian Scheinkman](https://github.com/SchSeba) | Red Hat | |
> +| [Marcin Franczyk](https://github.com/mfranczy) | Red Hat | |
> +| [Fabian Deutsch](https://github.com/fabiand) | Red Hat | |
> +| [Stu Gott](https://github.com/stu-gott) | Red Hat | |
>
> Thanks for showing the openess, Vasily. We'll support you and are looking
> forward to your thoughts and opinions.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubevirt/community/pull/123#pullrequestreview-744254106>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAS5KZYFURDYC6G7ZZH7HL3T7ZU5BANCNFSM5ABWFKCA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

fabiand commented:
/approved

fabiand commented:
/lgtm

rthallisey commented:
Oh I didn't know I needed to reply in comment.  I gave a +1 to being added

vladikr commented:
/lgtm

davidvossel reviewed (commented):
/lgtm

Thanks!

rmohr reviewed (approved):
/lgtm

rmohr commented:
/approve

3 lgtm from core maintainers.

fabiand commented:
/hold cancel

fabiand commented:
/unhold
/remove-hold
